src/foam/nanos/notification/email/DAOResourceLoader.java

```diff
@@ -50,31 +54,26 @@ public static EmailTemplate findTemplate(X x, String name, String groupId, Strin
     */
 
     do {
-      EmailTemplate emailTemplate = (EmailTemplate) emailTemplateDAO
-        .find(
-          AND(
-            EQ(EmailTemplate.NAME,   name),
-            EQ(EmailTemplate.GROUP,  SafetyUtil.isEmpty(groupId) ? "*" : groupId),
-            EQ(EmailTemplate.LOCALE, locale)
-          ));
-
-      if ( emailTemplate == null ) {
-        emailTemplate = (EmailTemplate) emailTemplateDAO
-          .find(
-            AND(
-              EQ(EmailTemplate.NAME,  name),
-              EQ(EmailTemplate.GROUP, SafetyUtil.isEmpty(groupId) ? "*" : groupId)
-            ));
-      }
-
-      if ( emailTemplate == null && ! SafetyUtil.isEmpty(groupId) ) {
-        emailTemplate = (EmailTemplate) emailTemplateDAO
-          .find(
-            AND(
-              EQ(EmailTemplate.NAME,  name),
-              EQ(EmailTemplate.GROUP, "*")
-            ));
-      }
+      boolean group_  = ! SafetyUtil.isEmpty(groupId);
```

Comment:
Considering the do{}while() will always return prior to going through parent groups... I propose this:
```
public static EmailTemplate findTemplate(X x, String name, String groupId, String locale, String spid, Map templateArgs) {
    DAO groupDAO = (DAO) x.get("groupDAO");
    DAO emailTemplateDAO = (DAO) x.get("localEmailTemplateDAO");
    EmailTemplate emailTemplate = null;

    /*
    name  group locale spid
      Y     Y     Y     Y
      Y     Y     Y     *
      Y     Y     *     Y
      Y     Y     *     *
      Y     *     *     Y
      Y     *     *     *
    */

    List<String> groupIdList = new ArrayList<>();
    while ( ! SafetyUtil.isEmpty(groupId) ) {
      groupIdList.add(groupId);
      if ( ! SafetyUtil.isEmpty(spid) )
        emailTemplate = findTemplateHelper(x, name, groupId, locale, spid, templateArgs);
      if ( emailTemplate != null ) return emailTemplate;
      // next group setup
      Group group = (Group) groupDAO.find(groupId);
      groupId = ( group != null && ! SafetyUtil.isEmpty(group.getParent()) ) ? group.getParent() : "";
    };
    // loop through grps again without spid
    for ( String group : groupIdList ) {
      emailTemplate = findTemplateHelper(x, name, group, locale, "", templateArgs);
      if ( emailTemplate != null )
        return emailTemplate;
    }
     // loop through grps again without local - default 'en'
    for ( String group : groupIdList ) {
      emailTemplate = findTemplateHelper(x, name, groupId, "en", spid, templateArgs);
      if ( emailTemplate != null )
        return emailTemplate;
    }
    // loop through grps again without local or spid
    for ( String group : groupIdList ) {
      emailTemplate = findTemplateHelper(x, name, groupId, "en", "", templateArgs);
      if ( emailTemplate != null )
        return emailTemplate;
    }
    // check spid no local or group
    emailTemplate = findTemplateHelper(x, name,"", "en", spid, templateArgs);
    // check just name - if not null
    if ( emailTemplate == null ) {
      emailTemplate = findTemplateHelper(x, name, "", "en", "", templateArgs);
    }
    
    return emailTemplate;
  }
  ```
 you'll need to import java.util.ArrayList ... also unfortunately have not tested, but think the logic shows the suggestion. Such that we will try the parents before filtering out one of the possible variables.

Reply:
updated